File: raspberry/src/mqtt_telemetry/storage/influxdb.py
"""
InfluxDB storage backend
Optimized for time-series data
"""

from influxdb_client import InfluxDBClient, Point, WritePrecision
from influxdb_client.client.write_api import SYNCHRONOUS
from typing import List, Dict, Any, Optional
from datetime import datetime
import logging

from .base import StorageBackend, StorageInfo

logger = logging.getLogger(__name__)


class InfluxDBStorage(StorageBackend):
    """InfluxDB storage backend for time-series data"""

    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        influx_config = config.get('influxdb', {})

        self.url = influx_config.get('url', 'http://localhost:8086')
        self.token = influx_config.get('token')
        self.org = influx_config.get('org', 'telemetry')
        self.bucket = influx_config.get('bucket', 'telemetry')
        self.batch_size = influx_config.get('batch_size', 1000)

        self.client: Optional[InfluxDBClient] = None
        self.write_api = None
        self.query_api = None

    async def initialize(self):
        """Initialize InfluxDB connection"""
        try:
            self.client = InfluxDBClient(
                url=self.url,
                token=self.token,
                org=self.org
            )

            self.write_api = self.client.write_api(write_options=SYNCHRONOUS)
            self.query_api = self.client.query_api()

            # Verify connection
            health = self.client.health()
            if health.status != "pass":
                raise Exception(f"InfluxDB health check failed: {health.message}")

            logger.info(f"InfluxDB storage initialized: {self.url}")

        except Exception as e:
            logger.error(f"Failed to initialize InfluxDB: {e}")
            raise

    async def store_message(self, message: Dict[str, Any]) -> bool:
        """Store single message"""
        try:
            point = self._message_to_point(message)
            self.write_api.write(bucket=self.bucket, org=self.org, record=point)
            return True

        except Exception as e:
            logger.error(f"Failed to store message: {e}")
            return False

    async def store_batch(self, messages: List[Dict[str, Any]]) -> bool:
        """Store multiple messages"""
        try:
            points = [self._message_to_point(msg) for msg in messages]
            self.write_api.write(bucket=self.bucket, org=self.org, record=points)

            logger.debug(f"Stored batch of {len(messages)} messages")
            return True

        except Exception as e:
            logger.error(f"Failed to store batch: {e}")
            return False

    def _message_to_point(self, message: Dict[str, Any]) -> Point:
        """Convert message to InfluxDB Point"""
        topic = message.get('topic', 'unknown')
        payload = message.get('payload', {})
        timestamp = message.get('timestamp', datetime.now())

        # Create point with measurement name from topic
        measurement = topic.replace('/', '_')
        point = Point(measurement).time(timestamp, WritePrecision.NS)

        # Add topic as tag
        point.tag("topic", topic)

        # Extract device_id or similar as tag if present
        if 'device_id' in payload:
            point.tag("device_id", str(payload['device_id']))

        if 'vehicle_id' in payload:
            point.tag("vehicle_id", str(payload['vehicle_id']))

        # Add payload fields
        for key, value in payload.items():
            if key not in ['device_id', 'vehicle_id']:  # Skip tags
                if isinstance(value, (int, float)):
                    point.field(key, float(value))
                elif isinstance(value, bool):
                    point.field(key, value)
                elif isinstance(value, str):
                    point.field(key, value)

        return point

    async def query(
        self,
        topic: Optional[str] = None,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None,
        limit: int = 1000
    ) -> List[Dict[str, Any]]:
        """Query messages from InfluxDB"""
        try:
            # Build Flux query
            query = f'from(bucket: "{self.bucket}")'

            # Time range
            if start_time:
                query += f' |> range(start: {start_time.isoformat()}Z'
                if end_time:
                    query += f', stop: {end_time.isoformat()}Z'
                query += ')'
            else:
                query += ' |> range(start: -7d)'  # Default last 7 days

            # Filter by topic if specified
            if topic:
                measurement = topic.replace('/', '_')
                query += f' |> filter(fn: (r) => r._measurement == "{measurement}")'

            # Limit
            query += f' |> limit(n: {limit})'

            # Execute query
            tables = self.query_api.query(query, org=self.org)

            # Convert results
            messages = []
            for table in tables:
                for record in table.records:
                    messages.append({
                        'topic': record.values.get('topic'),
                        'timestamp': record.get_time(),
                        'field': record.get_field(),
                        'value': record.get_value()
                    })

            return messages

        except Exception as e:
            logger.error(f"Query error: {e}")
            return []

    async def get_info(self) -> StorageInfo:
        """Get storage information"""
        try:
            # Get bucket info
            buckets_api = self.client.buckets_api()
            bucket = buckets_api.find_bucket_by_name(self.bucket)

            # Count approximate messages
            query = f'''
                from(bucket: "{self.bucket}")
                  |> range(start: 0)
                  |> count()
            '''

            tables = self.query_api.query(query, org=self.org)
            total_messages = sum(
                record.get_value()
                for table in tables
                for record in table.records
            )

            return StorageInfo(
                backend_type="influxdb",
                total_messages=total_messages,
                total_size_bytes=0,  # InfluxDB doesn't easily expose this
                oldest_message=None,
                newest_message=None
            )

        except Exception as e:
            logger.error(f"Error getting storage info: {e}")
            return StorageInfo(
                backend_type="influxdb",
                total_messages=0,
                total_size_bytes=0
            )

    async def cleanup(self, before: datetime) -> int:
        """Delete messages before timestamp"""
        try:
            # Use InfluxDB delete API
            delete_api = self.client.delete_api()

            predicate = f'_time < {before.isoformat()}Z'

            delete_api.delete(
                start=datetime(1970, 1, 1),
                stop=before,
                predicate=predicate,
                bucket=self.bucket,
                org=self.org
            )

            logger.info(f"Deleted messages before {before}")
            return 0  # InfluxDB doesn't return count

        except Exception as e:
            logger.error(f"Cleanup error: {e}")
            return 0

    async def close(self):
        """Close InfluxDB connection"""
        if self.client:
            self.client.close()
            logger.info("InfluxDB storage closed")

File: server/src/database/backends/influxdb.ts
/**
 * InfluxDB Database Backend
 * Optimized for time-series telemetry data
 */

import { InfluxDB, Point, WriteApi, QueryApi, HttpError } from '@influxdata/influxdb-client';
import { DatabaseBackend, QueryOptions } from '../manager';
import { DatabaseConfig } from '../../config';
import { logger } from '../../utils/logger';

export class InfluxDBBackend implements DatabaseBackend {
  private client!: InfluxDB;
  private writeApi!: WriteApi;
  private queryApi!: QueryApi;
  private bucket!: string;
  private org!: string;

  constructor(private config: DatabaseConfig) {}

  async initialize(): Promise<void> {
    const influxConfig = this.config.influxdb;

    if (!influxConfig) {
      throw new Error('InfluxDB configuration not found');
    }

    const url = influxConfig.url || 'http://localhost:8086';
    const token = influxConfig.token;
    this.org = influxConfig.org || 'mqtt-telemetry';
    this.bucket = influxConfig.bucket || 'telemetry';

    if (!token) {
      throw new Error('InfluxDB token is required');
    }

    this.client = new InfluxDB({ url, token });
    this.writeApi = this.client.getWriteApi(this.org, this.bucket, 'ns');
    this.queryApi = this.client.getQueryApi(this.org);

    // Configure write options
    this.writeApi.useDefaultTags({ source: 'mqtt-telemetry-server' });

    logger.info(`InfluxDB backend initialized (url: ${url}, bucket: ${this.bucket})`);
  }

  async store(message: any): Promise<boolean> {
    try {
      const point = this.messageToPoint(message);
      this.writeApi.writePoint(point);

      // Flush to ensure write
      await this.writeApi.flush();

      return true;
    } catch (error) {
      logger.error('Failed to store message in InfluxDB', error);
      return false;
    }
  }

  async storeBatch(messages: any[]): Promise<boolean> {
    try {
      const points = messages.map(msg => this.messageToPoint(msg));
      this.writeApi.writePoints(points);

      // Flush to ensure all writes
      await this.writeApi.flush();

      return true;
    } catch (error) {
      logger.error('Failed to store batch in InfluxDB', error);
      return false;
    }
  }

  private messageToPoint(message: any): Point {
    const point = new Point('mqtt_message')
      .timestamp(message.timestamp)
      .tag('topic', message.topic);

    // Extract numeric fields from payload for efficient querying
    const payload = message.payload;

    if (typeof payload === 'object' && payload !== null) {
      // Add all numeric values as fields
      for (const [key, value] of Object.entries(payload)) {
        if (typeof value === 'number') {
          point.floatField(key, value);
        } else if (typeof value === 'boolean') {
          point.booleanField(key, value);
        } else if (typeof value === 'string') {
          point.stringField(key, value);
        }
      }

      // Store full payload as string for complex objects
      if (Object.keys(payload).some(k => typeof payload[k] === 'object')) {
        point.stringField('payload_json', JSON.stringify(payload));
      }
    } else {
      // Store primitive values directly
      point.stringField('payload', String(payload));
    }

    // Add received_at as a field for query purposes
    point.timestamp(message.timestamp);
    point.intField('received_at_ms', message.receivedAt.getTime());

    return point;
  }

  async query(options: QueryOptions): Promise<any[]> {
    try {
      let fluxQuery = `from(bucket: "${this.bucket}")`;

      // Time range
      if (options.startTime) {
        const startMs = options.startTime.getTime();
        fluxQuery += `\n  |> range(start: ${Math.floor(startMs / 1000)})`;
      } else {
        fluxQuery += `\n  |> range(start: -30d)`;
      }

      if (options.endTime) {
        const endMs = options.endTime.getTime();
        fluxQuery += `\n  |> range(stop: ${Math.floor(endMs / 1000)})`;
      }

      // Filter by measurement
      fluxQuery += `\n  |> filter(fn: (r) => r._measurement == "mqtt_message")`;

      // Filter by topic
      if (options.topic) {
        fluxQuery += `\n  |> filter(fn: (r) => r.topic == "${options.topic}")`;
      }

      // Sort by time descending
      fluxQuery += `\n  |> sort(columns: ["_time"], desc: true)`;

      // Limit
      if (options.limit) {
        fluxQuery += `\n  |> limit(n: ${options.limit})`;
      }

      logger.debug('InfluxDB query:', fluxQuery);

      const rows: any[] = [];
      const messagesMap = new Map<string, any>();

      await this.queryApi.queryRows(fluxQuery, {
        next: (row: string[], tableMeta: any) => {
          const record = tableMeta.toObject(row);

          const timestamp = new Date(record._time);
          const key = `${timestamp.getTime()}_${record.topic}`;

          if (!messagesMap.has(key)) {
            messagesMap.set(key, {
              topic: record.topic,
              payload: {},
              timestamp: timestamp,
              receivedAt: new Date(record.received_at_ms || record._time),
            });
          }

          const message = messagesMap.get(key);

          // Reconstruct payload from fields
          if (record._field === 'payload_json') {
            message.payload = JSON.parse(record._value);
          } else if (record._field !== 'received_at_ms') {
            message.payload[record._field] = record._value;
          }
        },
        error: (error: HttpError) => {
          logger.error('InfluxDB query error', error);
        },
        complete: () => {
          // Query completed
        },
      });

      // Convert map to array
      rows.push(...Array.from(messagesMap.values()));

      return rows;
    } catch (error) {
      logger.error('Failed to query InfluxDB', error);
      return [];
    }
  }

  async cleanup(before: Date): Promise<number> {
    try {
      const startMs = 0;
      const stopMs = Math.floor(before.getTime() / 1000);

      const deleteQuery = `
        from(bucket: "${this.bucket}")
          |> range(start: ${startMs}, stop: ${stopMs})
          |> filter(fn: (r) => r._measurement == "mqtt_message")
      `;

      // Note: InfluxDB delete requires a predicate
      // This is a simplified version - in production, use the Delete API
      logger.warn('InfluxDB cleanup not fully implemented - use retention policies instead');

      return 0;
    } catch (error) {
      logger.error('Failed to cleanup InfluxDB', error);
      return 0;
    }
  }

  async getStats(): Promise<any> {
    try {
      // Count total messages
      const countQuery = `
        from(bucket: "${this.bucket}")
          |> range(start: 0)
          |> filter(fn: (r) => r._measurement == "mqtt_message")
          |> count()
      `;

      let totalMessages = 0;

      await this.queryApi.queryRows(countQuery, {
        next: (row: string[], tableMeta: any) => {
          const record = tableMeta.toObject(row);
          totalMessages += record._value || 0;
        },
        error: (error: HttpError) => {
          logger.error('InfluxDB stats query error', error);
        },
        complete: () => {},
      });

      return {
        total_messages: totalMessages,
        backend: 'influxdb',
        bucket: this.bucket,
        org: this.org,
      };
    } catch (error) {
      logger.error('Failed to get InfluxDB stats', error);
      return null;
    }
  }

  async close(): Promise<void> {
    if (this.writeApi) {
      try {
        await this.writeApi.close();
        logger.info('InfluxDB write API closed');
      } catch (error) {
        logger.error('Error closing InfluxDB write API', error);
      }
    }
  }
}
